app/src/pages/playground/PlaygroundOutput.tsx

```diff
@@ -67,80 +104,367 @@ function PlaygroundOutputMessage({ message }: { message: ChatMessage }) {
+    throw new Error("We only support chat templates for now");
+  }
 
-  const OutputEl = useMemo(() => {
-    if (hasRunId) {
-      return (
-        <PlaygroundOutputText key={runId} playgroundInstanceId={instanceId} />
+  const [generateChatCompletion] = useMutation<PlaygroundOutputMutation>(
+    graphql`
+      mutation PlaygroundOutputMutation($input: ChatCompletionInput!) {
+        chatCompletion(input: $input) {
+          __typename
+          content
+          errorMessage
+          span {
+            id
+          }
+          toolCalls {
+            id
+            function {
+              name
+              arguments
+            }
+          }
+        }
+      }
+    `
+  );
+
+  const hasRunId = instance?.activeRunId != null;
+  const notifyError = useNotifyError();
+  const chatCompletionParams = useMemo(() => {
+    if (instance.template.__type !== "chat") {
+      throw new Error("We only support chat templates for now");
+    }
+    let invocationParameters: InvocationParameterInput[] = [
+      ...instance.model.invocationParameters,
+    ];
+    if (instance.tools.length > 0) {
+      invocationParameters.push({
+        invocationName: TOOL_CHOICE_PARAM_NAME,
+        valueJson: instance.toolChoice,
+      });
+    } else {
+      invocationParameters = invocationParameters.filter(
+        (param) =>
+          param.invocationName !== TOOL_CHOICE_PARAM_NAME &&
+          param.canonicalName !== TOOL_CHOICE_PARAM_CANONICAL_NAME
       );
     }
-    if (isChatMessages(instance.output)) {
-      const messages = instance.output;
+    const azureModelParams =
+      instance.model.provider === "AZURE_OPENAI"
+        ? {
+            endpoint: instance.model.endpoint,
+            apiVersion: instance.model.apiVersion,
+          }
+        : {};
+
+    return {
+      messages: instance.template.messages.map(toGqlChatCompletionMessage),
+      model: {
+        providerKey: instance.model.provider,
+        name: instance.model.modelName || "",
+        ...azureModelParams,
+      },
+      invocationParameters: invocationParameters,
+      template: {
+        variables: templateVariables,
+        language: templateLanguage,
+      },
+      tools: instance.tools.length
+        ? instance.tools.map((tool) => tool.definition)
+        : undefined,
+      apiKey: credentials[instance.model.provider],
+    };
+  }, [
+    credentials,
+    instance.model.apiVersion,
+    instance.model.endpoint,
+    instance.model.invocationParameters,
+    instance.model.modelName,
+    instance.model.provider,
+    instance.template.__type,
+    instance.template.messages,
+    instance.toolChoice,
+    instance.tools,
+    templateLanguage,
+    templateVariables,
+  ]);
+
+  const [outputContent, setOutputContent] = useState<OutputContent>(
+    instance.output
+  );
+  const [toolCalls, setToolCalls] = useState<readonly PartialOutputToolCall[]>(
+    []
+  );
+
+  const onNext = useCallback(
```

Comment:
tried to pull this out of the component cause it's very hard to read with how much is oging on in here but it's so coupled with somany pieces of state, it was hard to get the setters and notifications all plumbed through